### PR TITLE
Add CPP Client

### PR DESCRIPTION
**Pull request by leszko**
We need to wait with this PR after this fix is released: https://github.com/hazelcast/hazelcast-cpp-client/pull/740.

@ihsandemir could I ask you to:
1. Read the guide and check if everything is correct.
2. Think if we can get rid of this `config.get_network_config().set_connection_attempt_limit(INT32_MAX);`. The default value `2` seems low. Maybe we can increase it starting from the `4.0` version?
3. You mentioned we can improve Dockerfile with cmake, could you do it?
Thanks!

### CODE REVIEW

File: doc/modules/ROOT/pages/index.adoc [hunk 1]
```diff
@@ -158,6 +167,20 @@ const client = await Client.newHazelcastClient(clientConfig);
 ----
 --
 
+C++::
++
+--
+[source, cpp]
+----
+hazelcast::client::client_config config;
+config.get_network_config().add_address(
```
Comment (ihsandemir) [inline doc/modules/ROOT/pages/index.adoc:176]: can make less lines like this: 
```C++
                hazelcast::client::client_config config;
                config.get_network_config().add_address({"localhost", 5701}).set_connection_attempt_limit(INT32_MAX);
                hazelcast::client::hazelcast_client hz(std::move(config));

```

File: cpp/Client.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#define BOOST_THREAD_VERSION 5
```
Comment (ihsandemir) [inline cpp/Client.cpp:1]: can be provided in g++ compile line as `-DBOOST_THREAD_VERSION=5`

**Comment by ihsandemir** [comment]
> We need to wait with this PR after this fix is released: [hazelcast/hazelcast-cpp-client#740](https://github.com/hazelcast/hazelcast-cpp-client/pull/740).
> 
> @ihsandemir could I ask you to:
> 
> 1. Read the guide and check if everything is correct.
> 2. Think if we can get rid of this `config.get_network_config().set_connection_attempt_limit(INT32_MAX);`. The default value `2` seems low. Maybe we can increase it starting from the `4.0` version?
> 3. You mentioned we can improve Dockerfile with cmake, could you do it?
>    Thanks!

2. We can change the value but what will be the default value? Java client also changed it to a new config `ConnectionRetryConfig` which has a timeout of 20 seconds, is this enough time?
3. Sure, I can do that. It can be changed after this PR https://github.com/hazelcast/hazelcast-cpp-client/pull/723 is merged.

File: doc/modules/ROOT/pages/index.adoc (hunk 1 above)
Comment (leszko) [reply doc/modules/ROOT/pages/index.adoc]: Changed

File: cpp/Client.cpp (hunk 2 above)
Comment (leszko) [reply cpp/Client.cpp]: Changed

**Comment by leszko** [comment]
> > 2. Think if we can get rid of this `config.get_network_config().set_connection_attempt_limit(INT32_MAX);`. The default value `2` seems low. Maybe we can increase it starting from the `4.0` version?
> 1. We can change the value but what will be the default value? Java client also changed it to a new config `ConnectionRetryConfig` which has a timeout of 20 seconds, is this enough time?

@sancar do we plan to increase the default connection timeout in Java? Otherwise, we need to advise all k8 users to set it explicitly, which makes the configuration harder.

File: doc/modules/ROOT/pages/index.adoc [hunk 3]
```diff
@@ -182,6 +192,21 @@ client = hazelcast.HazelcastClient(
...
++
+--
+[source, cpp]
+----
+#include <hazelcast/client/hazelcast_client.h>
+
+hazelcast::client::client_config config;
+config.get_network_config()
+      .add_address({"hz-hazelcast", 5701})
+      .set_connection_attempt_limit(INT32_MAX);
```
Comment (ihsandemir) [inline doc/modules/ROOT/pages/index.adoc:205]:
```suggestion
```

File: doc/modules/ROOT/pages/index.adoc [hunk 4]
```diff
@@ -182,6 +192,21 @@ client = hazelcast.HazelcastClient(
...
+hazelcast::client::hazelcast_client hz(std::move(config));
```
Comment (ihsandemir) [inline doc/modules/ROOT/pages/index.adoc:206]:
```suggestion
auto hz = hazelcast::new_client(std::move(config)).get();
```

**Review by ihsandemir — CHANGES_REQUESTED** [review]

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**